cpp/src/preprocessor.cpp

```diff
@@ -36,70 +27,74 @@ std::map<int, int> PreProcessor::get_source_map() {
 /**
  * PreProcess text before it's parsed by context-free PEG grammar
  */
-void PreProcessor::process(std::vector<std::string> lines) {
-  // Append newline character to the end of each line
-  for (std::string &line : lines) {
-    line = line + "\n";
-  }
-
+std::string PreProcessor::process(std::vector<std::string> lines) {
   int new_indent;
 
-  for (int line_number = 0; line_number < lines.size(); line_number++) {
+  // Append newline character to the end of each line
+  std::transform(lines.begin(), lines.end(), lines.begin(), [](std::string line) { return line + "\n"; });
+
+  for (unsigned int line_number = 0; line_number < lines.size(); line_number++) {
     std::string &line = lines[line_number];
 
     if (in_literal) {
-      // To accommodate empty lines in multiline literals which are part of the literal
-      // TODO: link to test case
-      if (line.substr(0, line.length() - 1).empty()) {
-        new_indent = current_indent;
-        line = std::string(current_indent, ' ') + "\n";
-      } else {
-        new_indent = line.length() - boost::trim_left_copy(line).length();
-      }
+      process_line_in_literal(line, new_indent);
     } else {
       if (boost::trim_left_copy(line).empty()) continue;
       new_indent = line.length() - boost::trim_left_copy(line).length();
     }
 
     check_new_indent(new_indent);
     output += remove_indent_whitespace(line);
-    source_map[count(output.begin(), output.end(), '\n')] = line_number+1;
+    source_map[count(output.begin(), output.end(), '\n')] = line_number + 1;
     check_and_enter_literal(line);
   }
   close_tags(0);
+  return output;
+}
+
+/**
+ * To accommodate empty lines in multiline literals which are part of the literal
+ */
+void PreProcessor::process_line_in_literal(std::string& line, int& new_indent) {
```

Comment:
Maybe either that or a more descriptive comment. e.g.: Normally, we just discard blank lines. But in literals, we want to preserve spacing, but also not consider it a dedent